From 7762456c5b6eb6ec50426c171f87f700fe1fed9e Mon Sep 17 00:00:00 2001
From: Miroslav Lisik <mlisik@redhat.com>
Date: Tue, 20 May 2025 16:34:18 +0200
Subject: [PATCH 2/3] support for query limits in rack

---
 pcsd/conf/pcsd | 6 ++++++
 pcsd/pcsd.rb   | 5 +++++
 2 files changed, 11 insertions(+)

diff --git a/pcsd/conf/pcsd b/pcsd/conf/pcsd
index 0ffbd616..7206e95a 100644
--- a/pcsd/conf/pcsd
+++ b/pcsd/conf/pcsd
@@ -48,5 +48,11 @@ PCSD_SESSION_LIFETIME=3600
 # is 50 (even if set lower).
 PCSD_RESTART_AFTER_REQUESTS=200
 
+# These environment variables set the maximum query string bytesize and the
+# maximum number of query parameters that pcsd will attempt to parse.
+# See CVE-2025-46727 for details.
+#RACK_QUERY_PARSER_BYTESIZE_LIMIT=4194304
+#RACK_QUERY_PARSER_PARAMS_LIMIT=4096
+
 # Do not change
 RACK_ENV=production
diff --git a/pcsd/pcsd.rb b/pcsd/pcsd.rb
index 5b8d0a11..a62434c6 100644
--- a/pcsd/pcsd.rb
+++ b/pcsd/pcsd.rb
@@ -74,6 +74,11 @@ configure do
   CAPABILITIES_PCSD = capabilities_pcsd.freeze
 end
 
+error Rack::QueryParser::QueryLimitError do
+  $logger.warn(env['sinatra.error'].message)
+  return 400, env['sinatra.error'].message
+end
+
 def run_cfgsync
   node_connected = true
   if Cfgsync::ConfigSyncControl.sync_thread_allowed?()
-- 
2.49.0

